Source/DafnyCore/AST/Formatting.dfy

```diff
@@ -0,0 +1,176 @@
+      }
+    }
+
+    lemma TokenNextIsNullImpliesLastToken(middle: IToken, i: int)
+      requires middle.Valid() && this.Valid()
+      requires 0 <= i < |allTokens|
+      requires middle == allTokens[i]
+      requires middle.Next == null
+      ensures i == |allTokens|-1
+    {
+      if Next == null {
+      } else {
+        Next.TokenNextIsNullImpliesLastToken(middle, i-1);
+      }
+    }
+  }
+  class {:extern "HelperString"} {:compile false} HelperString {
+    static predicate FinishesByNewline(input: CsString)
+  }
+}
+module {:extern "Microsoft"} {:options "-functionSyntax:4"}  Microsoft {
+  module {:extern "Dafny"} Dafny {
+    module {:extern "Formatting"} Formatting {
+      import opened MicrosoftDafny
+      import opened System
+
+      const {:extern "System", "String.Empty"} CsStringEmpty: CsString;
+
+      trait IIndentationFormatter {
+        function Reindent(token: IToken, trailingTrivia: bool, precededByNewline: bool, indentation: CsString, lastIndentation: CsString): CsString
+        method GetIndentation(token: IToken, currentIndentation: CsString)
+          returns (
+            indentationBefore: CsString,
+            indentationBeforeSet: bool,
+            lastIndentation: CsString,
+            lastIndentationSet: bool,
+            indentationAfter: CsString,
+            indentationAfterSet: bool)
+          requires token.Valid()
+          ensures token.allTokens == old(token.allTokens)
+      }
+
+      lemma IsAllocated<T>(x: T)
+        ensures allocated(x) {
+      }
+
+      /** Prints the entire program while fixing identation, based on
+          1) indentation information provided by the IIndentationFormatter reindent
+          2) Reindentation algorithm provided by the same reindent */
+      method printSourceReindent(firstToken: IToken, reindent: IIndentationFormatter) returns (s: CsString)
+        requires firstToken.Valid()
+        ensures forall token <- firstToken.allTokens :: s.Contains(token.val)
+      {
+        var token: IToken? := firstToken;
+        var sb := new CsStringBuilder();
+        var currentIndent := CsStringEmpty;
+        ghost var i := 0;
+        var leadingTriviaWasPreceededByNewline := true;
+        var allTokens := firstToken.allTokens;
+        while(token != null)
+          decreases if token == null then 0 else |token.allTokens|
+          invariant 0 <= i <= |allTokens|
+          invariant if token == null then i == |allTokens| else
+                    && token.Valid()
+                    && i < |allTokens|
+                    && token == allTokens[i]
+          invariant forall t <- allTokens[0..i] :: sb.built.Contains(t.val)
+        {
+          if(token.Next == null) {
+            firstToken.TokenNextIsNullImpliesLastToken(token, i);
+          } else {
+            firstToken.TokenNextIsIPlus1(token, i);
+          }
+          IsAllocated(allTokens[0..i]);
+          var indentationBefore, indentationBeforeSet,
```

Comment:
Some return values from `GetIndentation` are unprocessed and immediately passed back to `Reindent`. That seems like an API design smell. Why do you need to first call `GetIndentation` and then call `Reindent`, can't it be a single call?

Reply:
I would answer that it's not an API, these are just internal calls, which, if we determine it's best for maintenance, could always be refactored later without changing Dafny's API. Whether these calls happen in Dafny or C# should not make a difference.
My preference goes towards putting as much code in Dafny as possible, and have the external APIs do one simple thing at a time, so unless you have a strong feeling about that and tell me more about this "smell", I prefer to keep these three calls in Dafny.

Reply:
> I would answer that it's not an API, these are just internal calls

I meant the internal API of the interface `IIndentationFormatter`. Does API imply external? Is `interface signature` a better term?

> Whether these calls happen in Dafny or C# should not make a difference.

I agree but if you have an interface then it's a bad API design if one method needs to be fed the output from another one. It's like:

```
interface X {
  SomeType StepOne()
  Result StepTwo(SomeType value);
}
```

versus

```
interface X {
  Result Compute();
}
```

Reply:
> > I would answer that it's not an API, these are just internal calls
> 
> I meant the internal API of the interface `IIndentationFormatter`. Does API imply external?

Yes it does. It's [Application Programming Interface](https://en.wikipedia.org/wiki/API), i.e. two executable speaking to each other. 

> Is `interface signature` a better term?

Yes it is in your case.

> > Whether these calls happen in Dafny or C# should not make a difference.
> 
> I agree but if you have an interface then it's a clumsy API if one method needs to be fed the output from another one. It's like:
> 
> ```
> interface X {
>   SomeType StepOne()
>   Result StepTwo(SomeType value);
> }
> ```
> 
> versus
> 
> ```
> interface X {
>   Result Compute();
> }
> ```

Your example is too caricature for me but makes me understand how you feel.
My inner value was that I want to write as much code in Dafny as possible, and summarizing it with only one seemed to contradicts my goal.

But I found a workaround ! I gathered the four method calls into a single one in the Dafny trait, that the C# interface inherited, so that that code is in Dafny as well.